src/stransport_stream.c

```diff
@@ -110,20 +110,38 @@ int stransport_certificate(git_cert **out, git_stream *stream)
 	return 0;
 }
 
+/*
+ * Contrary to typical network IO callbacks, Secure Transport write callback is
+ * expected to write *all* passed data, not just as much as it can, and any
+ * other case would be considered a failure.
+ *
+ * This behavior is actually not specified in the Apple documentation, but is
+ * required for things to work correctly (and incidentally, that's also how
+ * Apple implements it in its projects at opensource.apple.com).
+ */
 static OSStatus write_cb(SSLConnectionRef conn, const void *data, size_t *len)
 {
 	git_stream *io = (git_stream *) conn;
+	OSStatus error = noErr;
+	size_t off = 0;
 	ssize_t ret;
 
-	ret = git_stream_write(io, data, *len, 0);
-	if (ret < 0) {
-		*len = 0;
-		return -1;
-	}
+	do {
+		ret = git_stream_write(io, data + off, *len - off, 0);
+		if (ret < 0) {
+			error = -36; /* "ioErr" from MacErrors.h */
+			break;
+		}
+		if (ret == 0) {
+			error = errSSLClosedGraceful;
+			break;
+		}
 
-	*len = ret;
+		off += ret;
+	} while (off < *len);
```

Comment:
Again, this is already the behaviour we expect from our streams, so there is absolutely no need to loop over it. When `git_stream_write()` returns, it will have written all of the data it was given, or it will return an error.